File: frontend/src/api/sub2api.ts
import request from '@/utils/request'
import type {
  PublicSub2APIHomeRequest,
  PublicSub2APIHomeResponse,
  GetSub2APIConfigRequest,
  GetSub2APIConfigResponse,
  UpdateSub2APIConfigRequest,
  UpdateSub2APIConfigResponse,
  TestSub2APIConnectionRequest,
  TestSub2APIConnectionResponse,
  SyncSub2APIUsageRequest,
  SyncSub2APIUsageResponse,
  GetSub2APISnapshotRequest,
  GetSub2APISnapshotResponse,
  GetSub2APIStatsRequest,
  GetSub2APIStatsResponse,
  GetSub2APIAdminTotalsRequest,
  GetSub2APIAdminTotalsResponse,
  GetSub2APIAdminTrendRequest,
  GetSub2APIAdminTrendResponse,
  GetSub2APIAdminTopRequest,
  GetSub2APIAdminTopResponse,
  GetSub2APIRecentRequestsRequest,
  GetSub2APIRecentRequestsResponse,
  ListSub2APIAnnouncementsResponse,
  CreateSub2APIAnnouncementRequest,
  UpdateSub2APIAnnouncementRequest,
  Sub2APIAnnouncementResponse,
  DeleteSub2APIAnnouncementResponse,
  ListSub2APITimelineResponse,
  CreateSub2APITimelineItemRequest,
  UpdateSub2APITimelineItemRequest,
  Sub2APITimelineItemResponse,
  DeleteSub2APITimelineItemResponse,
} from '@/types/v1/sub2api'

// 公开首页（无需鉴权）
export const getPublicSub2APIHome = (params: PublicSub2APIHomeRequest = {}) => {
  return request.get<PublicSub2APIHomeResponse>('/public/sub2api/home', { params })
}

// 配置
export const getSub2APIConfig = (params: GetSub2APIConfigRequest = {}) => {
  return request.get<GetSub2APIConfigResponse>('/sub2api/config', { params })
}

export const updateSub2APIConfig = (data: UpdateSub2APIConfigRequest) => {
  return request.put<UpdateSub2APIConfigResponse>('/sub2api/config', data)
}

export const testSub2APIConnection = (data: TestSub2APIConnectionRequest) => {
  return request.post<TestSub2APIConnectionResponse>('/sub2api/config/test', data)
}

// 用量
export const syncSub2APIUsage = (data: SyncSub2APIUsageRequest) => {
  return request.post<SyncSub2APIUsageResponse>('/sub2api/sync', data)
}

export const getSub2APISnapshot = (params: GetSub2APISnapshotRequest = {}) => {
  return request.get<GetSub2APISnapshotResponse>('/sub2api/snapshot', { params })
}

export const getPublicSub2APIStats = (params: GetSub2APIStatsRequest) => {
  return request.get<GetSub2APIStatsResponse>('/public/sub2api/stats', { params })
}

// 管理端用量汇总（标量指标 + 区间标签，需鉴权）
export const getSub2APIAdminTotals = (params: GetSub2APIAdminTotalsRequest) => {
  return request.get<GetSub2APIAdminTotalsResponse>('/sub2api/stats/totals', { params })
}

// 管理端用量趋势（按区间日内分桶/按天，需鉴权）
export const getSub2APIAdminTrend = (params: GetSub2APIAdminTrendRequest) => {
  return request.get<GetSub2APIAdminTrendResponse>('/sub2api/stats/trend', { params })
}

// 管理端用量排行（模型 + 分组，按 token 降序，需鉴权）
export const getSub2APIAdminTop = (params: GetSub2APIAdminTopRequest) => {
  return request.get<GetSub2APIAdminTopResponse>('/sub2api/stats/top', { params })
}

// 管理端最近请求（按时间区间分页，需鉴权）
export const getSub2APIRecentRequests = (params: GetSub2APIRecentRequestsRequest) => {
  return request.get<GetSub2APIRecentRequestsResponse>('/sub2api/recent-requests', { params })
}

// 公告
export const listSub2APIAnnouncements = () => {
  return request.get<ListSub2APIAnnouncementsResponse>('/sub2api/announcements')
}

export const createSub2APIAnnouncement = (data: CreateSub2APIAnnouncementRequest) => {
  return request.post<Sub2APIAnnouncementResponse>('/sub2api/announcements', data)
}

export const updateSub2APIAnnouncement = (data: UpdateSub2APIAnnouncementRequest) => {
  return request.put<Sub2APIAnnouncementResponse>(`/sub2api/announcements/${data.id}`, data)
}

export const deleteSub2APIAnnouncement = (id: string) => {
  return request.delete<DeleteSub2APIAnnouncementResponse>(`/sub2api/announcements/${id}`)
}

// 时间线
export const listSub2APITimeline = () => {
  return request.get<ListSub2APITimelineResponse>('/sub2api/timeline')
}

export const createSub2APITimelineItem = (data: CreateSub2APITimelineItemRequest) => {
  return request.post<Sub2APITimelineItemResponse>('/sub2api/timeline', data)
}

export const updateSub2APITimelineItem = (data: UpdateSub2APITimelineItemRequest) => {
  return request.put<Sub2APITimelineItemResponse>(`/sub2api/timeline/${data.id}`, data)
}

export const deleteSub2APITimelineItem = (id: string) => {
  return request.delete<DeleteSub2APITimelineItemResponse>(`/sub2api/timeline/${id}`)
}

File: internal/service/sub2api.go
package service

import (
	"context"
	"time"

	v1 "momoko/api/gen/v1"
	"momoko/internal/biz"
)

type Sub2APIService struct {
	v1.UnimplementedSub2APIManagerServer

	uc *biz.Sub2APIUsecase
}

func NewSub2APIService(uc *biz.Sub2APIUsecase) *Sub2APIService {
	return &Sub2APIService{uc: uc}
}

func (s *Sub2APIService) PublicSub2APIHome(ctx context.Context, _ *v1.PublicSub2APIHomeRequest) (*v1.PublicSub2APIHomeResponse, error) {
	home, err := s.uc.PublicHome(ctx)
	if err != nil {
		return nil, err
	}
	return &v1.PublicSub2APIHomeResponse{Home: home}, nil
}

func (s *Sub2APIService) GetSub2APIConfig(ctx context.Context, _ *v1.GetSub2APIConfigRequest) (*v1.GetSub2APIConfigResponse, error) {
	config, err := s.uc.Config(ctx)
	if err != nil {
		return nil, err
	}
	return &v1.GetSub2APIConfigResponse{Config: config}, nil
}

func (s *Sub2APIService) UpdateSub2APIConfig(ctx context.Context, req *v1.UpdateSub2APIConfigRequest) (*v1.UpdateSub2APIConfigResponse, error) {
	config, err := s.uc.UpdateConfig(ctx, req)
	if err != nil {
		return nil, err
	}
	return &v1.UpdateSub2APIConfigResponse{Config: config}, nil
}

func (s *Sub2APIService) TestSub2APIConnection(ctx context.Context, req *v1.TestSub2APIConnectionRequest) (*v1.TestSub2APIConnectionResponse, error) {
	return s.uc.TestConnection(ctx, req)
}

func (s *Sub2APIService) SyncSub2APIUsage(ctx context.Context, req *v1.SyncSub2APIUsageRequest) (*v1.SyncSub2APIUsageResponse, error) {
	snapshot, err := s.uc.SyncUsage(ctx, req.GetFull())
	if err != nil {
		return nil, err
	}
	return &v1.SyncSub2APIUsageResponse{Snapshot: snapshot}, nil
}

func (s *Sub2APIService) GetSub2APISnapshot(ctx context.Context, _ *v1.GetSub2APISnapshotRequest) (*v1.GetSub2APISnapshotResponse, error) {
	snapshot, err := s.uc.Snapshot(ctx)
	if err != nil {
		return nil, err
	}
	return &v1.GetSub2APISnapshotResponse{Snapshot: snapshot}, nil
}

func (s *Sub2APIService) GetPublicSub2APIStats(ctx context.Context, req *v1.GetSub2APIStatsRequest) (*v1.GetSub2APIStatsResponse, error) {
	stats, err := s.uc.PublicStats(ctx, req.GetRangeDays())
	if err != nil {
		return nil, err
	}
	return &v1.GetSub2APIStatsResponse{Stats: stats}, nil
}

func (s *Sub2APIService) GetSub2APIStats(ctx context.Context, req *v1.GetSub2APIAdminStatsRequest) (*v1.GetSub2APIAdminStatsResponse, error) {
	start, end := rangeFromMillis(req.GetStartTime(), req.GetEndTime())
	stats, err := s.uc.AdminStats(ctx, start, end)
	if err != nil {
		return nil, err
	}
	return &v1.GetSub2APIAdminStatsResponse{Stats: stats}, nil
}

func (s *Sub2APIService) GetSub2APIRecentRequests(ctx context.Context, req *v1.GetSub2APIRecentRequestsRequest) (*v1.GetSub2APIRecentRequestsResponse, error) {
	start, end := rangeFromMillis(req.GetStartTime(), req.GetEndTime())
	list, total, err := s.uc.RecentRequests(ctx, start, end, int(req.GetPage()), int(req.GetPageSize()))
	if err != nil {
		return nil, err
	}
	return &v1.GetSub2APIRecentRequestsResponse{
		RecentRequests: list,
		Total:          int64(total),
		Page:           req.GetPage(),
		PageSize:       req.GetPageSize(),
	}, nil
}

// rangeFromMillis 将 Unix 毫秒（<=0 视为未指定）转为时间段端点，归一化交由 usecase/service 处理。
func rangeFromMillis(startMs, endMs int64) (time.Time, time.Time) {
	var start, end time.Time
	if startMs > 0 {
		start = time.UnixMilli(startMs)
	}
	if endMs > 0 {
		end = time.UnixMilli(endMs)
	}
	return start, end
}

func (s *Sub2APIService) ListSub2APIAnnouncements(ctx context.Context, _ *v1.ListSub2APIAnnouncementsRequest) (*v1.ListSub2APIAnnouncementsResponse, error) {
	list, err := s.uc.ListAnnouncements(ctx)
	if err != nil {
		return nil, err
	}
	return &v1.ListSub2APIAnnouncementsResponse{Announcements: list}, nil
}

func (s *Sub2APIService) CreateSub2APIAnnouncement(ctx context.Context, req *v1.CreateSub2APIAnnouncementRequest) (*v1.Sub2APIAnnouncementResponse, error) {
	item, err := s.uc.CreateAnnouncement(ctx, req)
	if err != nil {
		return nil, err
	}
	return &v1.Sub2APIAnnouncementResponse{Announcement: item}, nil
}

func (s *Sub2APIService) UpdateSub2APIAnnouncement(ctx context.Context, req *v1.UpdateSub2APIAnnouncementRequest) (*v1.Sub2APIAnnouncementResponse, error) {
	item, err := s.uc.UpdateAnnouncement(ctx, req)
	if err != nil {
		return nil, err
	}
	return &v1.Sub2APIAnnouncementResponse{Announcement: item}, nil
}

func (s *Sub2APIService) DeleteSub2APIAnnouncement(ctx context.Context, req *v1.DeleteSub2APIAnnouncementRequest) (*v1.DeleteSub2APIAnnouncementResponse, error) {
	if err := s.uc.DeleteAnnouncement(ctx, req.GetId()); err != nil {
		return nil, err
	}
	return &v1.DeleteSub2APIAnnouncementResponse{}, nil
}

func (s *Sub2APIService) ListSub2APITimeline(ctx context.Context, _ *v1.ListSub2APITimelineRequest) (*v1.ListSub2APITimelineResponse, error) {
	list, err := s.uc.ListTimeline(ctx)
	if err != nil {
		return nil, err
	}
	return &v1.ListSub2APITimelineResponse{Timeline: list}, nil
}

func (s *Sub2APIService) CreateSub2APITimelineItem(ctx context.Context, req *v1.CreateSub2APITimelineItemRequest) (*v1.Sub2APITimelineItemResponse, error) {
	item, err := s.uc.CreateTimelineItem(ctx, req)
	if err != nil {
		return nil, err
	}
	return &v1.Sub2APITimelineItemResponse{Item: item}, nil
}

func (s *Sub2APIService) UpdateSub2APITimelineItem(ctx context.Context, req *v1.UpdateSub2APITimelineItemRequest) (*v1.Sub2APITimelineItemResponse, error) {
	item, err := s.uc.UpdateTimelineItem(ctx, req)
	if err != nil {
		return nil, err
	}
	return &v1.Sub2APITimelineItemResponse{Item: item}, nil
}

func (s *Sub2APIService) DeleteSub2APITimelineItem(ctx context.Context, req *v1.DeleteSub2APITimelineItemRequest) (*v1.DeleteSub2APITimelineItemResponse, error) {
	if err := s.uc.DeleteTimelineItem(ctx, req.GetId()); err != nil {
		return nil, err
	}
	return &v1.DeleteSub2APITimelineItemResponse{}, nil
}
